manta-accounting/src/wallet/signer/mod.rs

```diff
@@ -835,6 +846,217 @@ where
+    Eq(bound = "Asset<C>: Eq, Identifier<C>: Eq"),
+    Hash(bound = "Asset<C>: Hash, Identifier<C>: Hash"),
+    PartialEq(bound = "Asset<C>: PartialEq, Identifier<C>: PartialEq")
+)]
+pub struct ConsolidationPrerequest<C>(pub Vec<IdentifiedAsset<C>>)
+where
+    C: transfer::Configuration + ?Sized;
+
+impl<C> ConsolidationPrerequest<C>
+where
+    C: transfer::Configuration + ?Sized,
+{
+    /// Builds a new [`ConsolidationPrerequest`] from `assets`.
+    #[inline]
+    pub fn new(assets: Vec<IdentifiedAsset<C>>) -> Self {
+        Self(assets)
+    }
+}
+
+/// Consolidation Request
+///
+/// # Note
+///
+/// A consolidation request is a vector of [`IdentifiedAsset`]s satisfying the
+/// following conditions:
+/// 1) It contains no duplicates.
+/// 2) It contains no assets with value zero.
+/// 3) It contains at least two assets.
+/// 4) All the assets contained in it share the same asset id.
+#[cfg_attr(
+    feature = "serde",
+    derive(Deserialize, Serialize),
+    serde(
+        bound(
+            deserialize = r"Asset<C>: Deserialize<'de>, 
+                Identifier<C>: Deserialize<'de>",
+            serialize = r"Asset<C>: Serialize, 
+                Identifier<C>: Serialize"
+        ),
+        crate = "manta_util::serde",
+        deny_unknown_fields
+    )
+)]
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = "Asset<C>: Clone, Identifier<C>: Clone"),
+    Debug(bound = "Asset<C>: Debug, Identifier<C>: Debug"),
+    Eq(bound = "Asset<C>: Eq, Identifier<C>: Eq"),
+    Hash(bound = "Asset<C>: Hash, Identifier<C>: Hash"),
+    PartialEq(bound = "Asset<C>: PartialEq, Identifier<C>: PartialEq")
+)]
+pub struct ConsolidationRequest<C>(Vec<IdentifiedAsset<C>>)
+where
+    C: transfer::Configuration + ?Sized;
+
+impl<C> ConsolidationRequest<C>
+where
+    C: transfer::Configuration,
+    IdentifiedAsset<C>: PartialEq,
+{
+    /// Builds a new [`ConsolidationRequest`] from `assets`.
+    #[inline]
+    fn new_unchecked(assets: Vec<IdentifiedAsset<C>>) -> Self {
+        Self(assets)
+    }
+
+    /// Builds a new [`ConsolidationRequest`] from `assets`, ensuring that:
+    /// 1) There are at least two [`IdentifiedAsset`]s in `assets`.
+    /// 2) There are no duplicates in `assets`.
+    /// 3) All assets in `assets` have the same asset id.
+    /// 4) No asset in `assets` has zero value.
+    #[inline]
+    pub fn new(assets: Vec<IdentifiedAsset<C>>) -> Option<Self> {
+        let number_of_assets = assets.len();
+        if number_of_assets < 2 {
+            return None;
+        }
+        let asset_id = &assets
+            .first()
+            .expect("This cannot fail because of the check above.")
+            .asset
+            .id;
+        (!(1..number_of_assets).any(|i| {
+            assets[i..].contains(&assets[i - 1])
+                && assets[i].asset.id.ne(asset_id)
+                && assets[i].asset.value.eq(&Default::default())
+        }))
```

Comment:
This won't return None when only one of these conditions is met, which seems sus.

Reply:
Good catch